File: backend/src/main/java/lan/dk/podcastserver/service/ProcessService.java
package lan.dk.podcastserver.service;


import io.vavr.control.Try;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import static io.vavr.API.Try;
/**
 * Created by kevin on 25/01/2016 for Podcast Server
 */
@Slf4j
@Service
public class ProcessService {

    public ProcessBuilder newProcessBuilder(String... command) { return new ProcessBuilder(command); }

    public Try<Process> start(ProcessBuilder processBuilder) {
        return Try(processBuilder::start);
    }

    public int pidOf(Process p) {
        return Try(() -> p.getClass().getSimpleName())
            .filter(className -> className.contains("UNIXProcess"))
            .mapTry(className -> p.getClass().getDeclaredField("pid"))
            .andThenTry(c -> c.setAccessible(true))
            .mapTry(c -> c.getInt(p))
            .getOrElse(-1);
    }

    public Try<Integer> waitFor(Process process) {
        return Try(process::waitFor);
    }
}

File: backend/src/main/kotlin/com/github/davinkevin/podcastserver/service/ProcessService.kt
package com.github.davinkevin.podcastserver.service

import org.springframework.stereotype.Service
import java.util.*

/**
 * Created by kevin on 25/01/2016 for Podcast Server
 */
@Service
class ProcessService {

    fun newProcessBuilder(vararg command: String) = ProcessBuilder(*command)

    fun start(processBuilder: ProcessBuilder): Process = processBuilder.start()

    fun pidOf(p: Process) = p.pid()

    fun waitFor(process: Process): Result<Int> = Result.runCatching { process.waitFor() }
}
